eventlog/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2020 GRAIL, Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+package cloudwatch
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs/cloudwatchlogsiface"
+	"github.com/grailbio/base/config"
+	"github.com/grailbio/base/errors"
+	"github.com/grailbio/base/eventlog/internal/marshal"
+	"github.com/grailbio/base/log"
+)
+
+// maxBatchSize is the maximum batch size of the events that we send to CloudWatch Logs, "calculated
+// as the sum of all event messages in UTF-8, plus 26 bytes for each log event". See:
+// https://docs.aws.amazon.com/sdk-for-go/api/service/cloudwatchlogs/#CloudWatchLogs.PutLogEvents
+const maxBatchSize = 1048576
+
+// maxSingleMessageSize is the maximum size, in bytes, of a single message
+// string of CloudWatch Log event.
+const maxSingleMessageSize = maxBatchSize - 26
+
+// syncInterval is the maximum duration that we will wait before sending any
+// buffered messages to CloudWatch.
+const syncInterval = 1 * time.Second
+
+// eventBufferSize is size of the channel buffer used to process events. When
+// this buffer is full, new events are dropped.
+const eventBufferSize = 32768
+
+func init() {
```

Comment:
Is there any use in some sort of `close` or `onExit` style hook? I imagine we want to throw events whenever bad things happen. What's the best way to ensure we lose as little as possible?

Reply:
It does look like you're buffering, so curious about that.

Reply:
The semantics are similar to other streams. You have to `Close` it to make sure data is flushed/synced. In general, this is all best-effort, and I think the short buffer interval (1 second) should general make it good enough. If clients do want to do more, it's up to them to set up to call `Close` appropriately, e.g. at exit, in a signal handler, etc. I'm inclined to not worry about it for now.